### ISSUE TITLE
Add optimizer tests

### ISSUE THREAD

**Issue by t-ae** [tensorflow/swift-apis#696]
Currently there's no tests for optimizers.
It leads bugs like https://github.com/tensorflow/swift-apis/pull/683#issuecomment-589523022.

I wrote simple convergence tests but I'm not sure if they are worth adding.
https://github.com/t-ae/swift-apis/blob/optimizer-tests/Tests/TensorFlowTests/OptimizerTests.swift
Is there better way to test them?

**Comment by saeta** [comment]
Hi @t-ae ! That is roughly how I'd try and write a convergence test for the optimizers. Can you confirm that the test fails on Linux due to the keypath issue you identified? If so, I think it'd be great to get these tests added to the test suite! (Out of curiosity, how long do they take to run?)

**Comment by t-ae** [comment]
> Can you confirm that the test fails on Linux due to the keypath issue you identified?

Yes. `AdaMax` and `AMSGrad` fails only on Linux. They are using `recursivelyAllWritableKeyPaths`.

> Out of curiosity, how long do they take to run?

It takes 5 secs for these 8 tests in my environment.
Only `AdaGrad` is slow to converge with default parameter.
I found [`tf.keras.optimizers.Adagrad`](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adagrad)'s default learning rate is 0.001 whereas [`keras.optimizers.Adagrad`](https://keras.io/optimizers/#adagrad)'s default learnign rate is 0.01.
I don't know what this difference means but with leraning rate 0.01 it converges as fast as the others.

**Comment by vballoli** [comment]
Hi, in my previous PR https://github.com/tensorflow/swift-apis/pull/564  @dan-zheng suggested adding an OptimizerTest which I'll be submitting today for RAdam inspired from https://github.com/tensorflow/addons/blob/b4cea81897/tensorflow_addons/optimizers/rectified_adam_test.py. Hopefully that can serve as a starting point for all other optimizers

**Comment by t-ae** [comment]
Now my OptimizerTests.swift is merged. But it seems to fail to catch the problem https://github.com/tensorflow/swift-apis/pull/700#discussion_r385097412.
Numerical test @vballoli introduced in #700 is better than my convergence test.
I'll look into it after #700 is merged.

**Comment by dan-zheng** [comment]
Closing, as both optimizer convergence tests and numerical correctness tests have been added. Thanks!
